replica/database.cc

```diff
@@ -468,6 +468,48 @@ void backlog_controller::update_controller(float shares) {
     _scheduling_group.set_shares(shares);
 }
 
+constexpr backlog_controller::control_point compaction_controller::middle_point() {
+    return { 1.5, 100 };
+}
+constexpr backlog_controller::control_point compaction_controller::base_last_point() {
+    return { base_normalization_factor, 1000 };
+}
+
+const float compaction_controller::minimum_effective_max_shares() {
+    return middle_point().output * 2;
+}
+const float compaction_controller::default_max_shares() {
+    return base_last_point().output;
+}
+
+backlog_controller::control_point compaction_controller::last_control_point(float max_shares) {
+    max_shares = std::max(max_shares, minimum_effective_max_shares());
+
+    auto base_slope = slope(middle_point(), base_last_point());
```

Comment:
Since this function is used from the updater it needs to to use the current slope which may change in the next patch, not the base slope, no?

Only if the last control point isn't set it should use `base_last_point()`, otherwise it should use `last_point()` (which should be defined)

Reply:
we want to find a normalization factor (NF), that produces the same slope as with (default NF, default max shares). so the only variable is max shares. the slope is a constant using middle and base last points, which never change.

Reply:
But if one changes first the agressiveness factor and then max_shares then we should keep the existing slope, not revert to the default one

Reply:
I don't think it matters. We want to preserve the slope as with the default settings (ag 1.0, max shares 1k).

The formula for finding NF is:

NF = (MAX_SHARES - 100) / ((AG * DEFAULT_SLOPE) + 1.5)

where 

DEFAULT_SLOPE = (1k - 100) / (30 - 1.5)

If I change AG to 2.0 first, it means a new NF is calculated using current max shares and *new* AG as params.

If max shares changes later, then another NF is calculated using AG previously set (2.0) and *new* max shares as params.

As long as the controller nows about both current AG and max shares config, it's able to calculate correctly NF without knowing about previous states.

Reply:
I think the confusion will be clarified if you look at the last patch. Current AG is applied to the base slope ((1k - 100) / (30 - 1.5)) when calculating the new last control point, on setting update.